### PR TITLE
Fix hard-wired buffer limit in the JS API

### PR DESCRIPTION
**Pull request by dcodeIO**
As a follow-up to https://github.com/WebAssembly/binaryen/pull/1392, this PR adds a new variant of `BinaryenModuleWrite` to the C-API for use by the JS-API, fixing the 1MB pre-allocation limit:

* **BinaryenModuleAllocateAndWrite**

It differs in that it also implicitly allocates buffers for the output structures that must be freed manually. If `sourceMapUrl` is `NULL`, it skips generating a source map.

JS-API is:

* Module#**emitBinary**([sourceMapUrl])

returning either just the binary if `sourceMapUrl` is `undefined` (maintaining backwards compatibility) or both a `binary` and a `sourceMap`.

`Module#emitBinaryWithSourceMap` introduced in the previous PR is removed again.

### CODE REVIEW

File: src/binaryen-c.cpp [hunk 1]
```diff
@@ -2063,7 +2063,7 @@ void BinaryenSetDebugInfo(int on) {
     std::cout << "  BinaryenSetDebugInfo(" << on << ");\n";
   }
 
-  globalPassOptions.debugInfo = bool(on);
+  globalPassOptions.debugInfo = on ? true : false;
```
Comment (kripken) [inline src/binaryen-c.cpp]: was this necessary?

File: src/binaryen-c.h [hunk 2]
```diff
@@ -707,6 +707,19 @@ typedef struct BinaryenBufferSizes {
...
+typedef struct BinaryenModuleToBinaryResult {
+  void* binary;
+  size_t binaryBytes;
+  char* sourceMap;
+} BinaryenModuleToBinaryResult;
+
+// Serializes a module into binary form, optionally including its source map if
+// sourceMapUrl has been specified. Uses the currently set global debugInfo option.
+// Allocated buffers for both the output and the source map must be freed manually.
+BinaryenModuleToBinaryResult BinaryenModuleToBinary(BinaryenModuleRef module, const char* sourceMapUrl);
```
Comment (kripken) [inline src/binaryen-c.h]: I'm not too sure about the name `BinaryenModuleToBinary`. My inclination is it should be more similar to the existing `BinaryenModuleWrite` maybe with a suffix that indicates how it differs. `BinaryenModuleWriteExtended`? `BinaryenModuleWrite2`?

Alternatively, maybe `BinaryenModuleAllocateAndWrite`?

File: src/binaryen-c.h [hunk 3]
```diff
@@ -707,6 +707,19 @@ typedef struct BinaryenBufferSizes {
...
+// the source map must be freed manually.
+typedef struct BinaryenModuleToBinaryResult {
+  void* binary;
+  size_t binaryBytes;
+  char* sourceMap;
+} BinaryenModuleToBinaryResult;
+
+// Serializes a module into binary form, optionally including its source map if
+// sourceMapUrl has been specified. Uses the currently set global debugInfo option.
+// Allocated buffers for both the output and the source map must be freed manually.
```
Comment (kripken) [inline src/binaryen-c.h]: please specify `freed manually using free()` (as e.g. `delete` could be used in C++, but shouldn't).

File: src/binaryen-c.cpp (hunk 1 above)
Comment (dcodeIO) [reply src/binaryen-c.cpp]: MSVC emitted a warning there, converting an int to a bool. Not sure why.

File: src/binaryen-c.h (hunk 2 above)
Comment (dcodeIO) [reply src/binaryen-c.h]: BinaryenModuleAllocateAndWrite seems good

File: src/binaryen-c.cpp (hunk 1 above)
Comment (binji) [reply src/binaryen-c.cpp]: drive by comment: `on != 0` or `!!on` are the ways I usually see this written.

Comment (dcodeIO) [reply src/binaryen-c.cpp]: Thanks, `on != 0` is what I used now.

**Comment by kripken** [comment]
Thanks, changes look good!

How about if we make more of an effort to update `docs/binaryen.js.Markdown`? It would be good to add to there with each PR like this.

**Comment by dcodeIO** [comment]
Sure, if copying my other docs here isn't an option, I can try to keep this one up to date, too.

**Comment by kripken** [comment]
Oh, yeah, another PR is fine.

Remind me of your other docs, where are they? Maybe we can just link to those, depending on the content.

**Comment by dcodeIO** [comment]
I made [these](https://github.com/AssemblyScript/binaryen.js) initially to include typing information for the `binaryen` npm package, and later extended them with links to the specs etc..

**Comment by kripken** [comment]
Thanks. Those look pretty nice. I think we should focus on one set of docs in one place, as right now for example yours are missing `Module.set_local` (the alias with an `_`), while the ones in this repo are missing other stuff. I'm fine to go with yours, in which case we should make sure everything documented here is in yours, and we should remove the docs here and add a link in the readme for where to find the docs. Thoughts?

**Comment by dcodeIO** [comment]
> for example yours are missing Module.set_local

The aliases are not explicitly listed, but noted at the bottom of the variable accesses section:

>> Variable access methods above are also aliased in underscore notation, e.g., get_local.

but I should probably change that so it is more visible. Do you have any preference on the default notation there? `get_local` or `getLocal`?

> I'm fine to go with yours, in which case we should make sure everything documented here is in yours

The docs are actually based on yours (initially was the same file), but I might have stripped one or two details in the process.

> Thoughts?

I always feel a bit guilty by hosting another version of the docs, or the binaryen package on npm, because people might misunderstand that I am not the primary author of these. My repository merely exists because I rely on up to date JS builds with TypeScript definitions for AssemblyScript, so I figured that I automate that and publish it to npm in case others have that need as well. If you have an idea how we could provide that in another form, under your control, I am also fine with moving the entire thing. If that's not important to you, I can also continue to host it in the AssemblyScript org.

**Comment by dcodeIO** [comment]
Related: https://github.com/WebAssembly/binaryen/issues/1349

**Comment by kripken** [comment]
Probably `get_local` should be the default, since it matches the wast format? Less "JS-ey", but still seems slightly better to me.

If it's not too hard, we can just constantly sync those two files? Unless there is a difference in content that should exist?

**Comment by dcodeIO** [comment]
Alright, I just split the API from the README to make this easier: https://github.com/AssemblyScript/binaryen.js/blob/master/API.md

Theoretically, that file could also be pulled from there somehow.

**Comment by dcodeIO** [comment]
Or, actually, that file could be updated here directly, and I include it into my repo's README, which would certainly make more sense and guarantee that things are aligned. Only think I'd need is that the `<!-- BEGIN API -->` comment is retained.

**Comment by dcodeIO** [comment]
Or: We could maintain it within the wiki :) Decisions, decisions.

**Comment by kripken** [comment]
An advantage to not doing it in the wiki is that then it uses the same git history as the code itself, so people can easily see older versions of the docs.

**Comment by dcodeIO** [comment]
That's true,. I guess I just like how easy it is to keep it updated now, e.g., adding just a sentence or correcting a typo even without a PR. Your call :)

**Comment by kripken** [comment]
I guess it's fine either way. Long-term if we have many users ;) we may want to move it from the wiki, but the wiki sounds ok for now. Let's just add a link to the wiki and remove the copy that isn't in the wiki.

**Pull request merged**